File: python/pcs_api/cexceptions.py
# -*- coding: utf-8 -*-
#
# Copyright (c) 2014 Netheos (http://www.netheos.net)
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#


from __future__ import absolute_import, unicode_literals, print_function


class CStorageError(Exception):
    """Base class for all cloud storage errors.
    Such exceptions have an optional 'message' and 'cause' attribute."""
    def __init__(self, message, cause=None):
        super(CStorageError, self).__init__(message)
        self.cause = cause

    def __str__(self):
        ret = "%s(%s)" % (self.__class__, self.message)
        if self.cause:
            ret += " (caused by %r)" % (self.cause,)
        return ret


class CInvalidFileTypeError(CStorageError):
    """Raised when performing an operation on a folder when a blob is expected,
    or when operating on a blob and a folder is expected.
    Also raised when downloading provider special files (eg google drive native docs)."""

    def __init__(self, c_path, expected_blob, message=None):
        """:param c_path: the problematic path
        :param expected_blob: if True, a blob was expected but a folder was found.
                              if False, a folder was expected but a blob was found
        :param message: optional message"""
        if not message:
            message = 'Invalid file type at %r (expected %s)' % \
                      (c_path, 'blob' if expected_blob else 'folder')
        super(CInvalidFileTypeError, self).__init__(message)
        self.path = c_path
        self.expected_blob = expected_blob


class CRetriableError(CStorageError):
    """Raised by RequestInvoker validation method, when request
    has failed but should be retried.

    This class is only a marker ; the underlying root exception
    is given by the 'cause' attribute.
    The optional 'delay' specifies how much one should wait before retrying"""
    def __init__(self, cause, delay=None):
        super(CRetriableError, self).__init__(message=None, cause=cause)
        self.delay = delay

    def get_delay(self):
        return self.delay

    #def __str__(self):
    #    return "%s(%s)" % (self.__class__, self.cause)


class CFileNotFoundError(CStorageError):
    """File has not been found (sometimes consecutive to http 404 error)"""
    def __init__(self, message, c_path):
        super(CFileNotFoundError, self).__init__(message)
        self.path = c_path


class CHttpError(CStorageError):
    """Raised when providers server answers non OK answers"""
    def __init__(self, request_method,
                 request_path,
                 status_code, reason,
                 message=None):
        super(CHttpError, self).__init__(message)
        self.request_method = request_method
        self.request_path = request_path
        self.status_code = status_code
        self.reason = reason

    def __str__(self):
        ret = "%s(%d %s) %s %s" % (
            self.__class__.__name__, self.status_code, self.reason, self.request_method, self.request_path )
        if self.message:
            ret += ' msg=%s' % self.message
        return ret

class CAuthenticationError(CHttpError):
    """http 401 error"""
    pass

File: cpp/libpcs_api/src/model/c_exceptions.cc
/**
 * Copyright (c) 2014 Netheos (http://www.netheos.net)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <exception>

#include "boost/algorithm/string.hpp"
#include "boost/format.hpp"
#include "boost/exception/all.hpp"

#include "cpprest/http_msg.h"

#include "pcs_api/c_exceptions.h"

namespace pcs_api {

CStorageException::CStorageException() {
}

CStorageException::CStorageException(const std::string& message) :
    message_(message) {
}

CStorageException::CStorageException(const std::string& message,
                                     std::exception_ptr p_cause) :
    message_(message), p_cause_(p_cause) {
}

std::string CStorageException::ToString() const {
    std::ostringstream tmp;
    tmp << "CStorageException: " << what();
    return tmp.str();
}



CHttpException::CHttpException(const std::string& message,
                               int status,
                               const std::string& reason,
                               const std::string& method,
                               const std::string& url) :
        CStorageException(message),
        status_(status),
        reason_(reason),
        request_method_(method),
        url_(url) {
}

static std::string CHttpExceptionToString(const CHttpException& che,
                                          std::string class_name) {
    std::ostringstream tmp;
    tmp << boost::format("%s: %s %s [%i/%s] %s") %
            class_name % che.request_method() % che.url()
            % che.status() % che.reason() % che.what();
    return tmp.str();
}

std::string CHttpException::ToString() const {
    return CHttpExceptionToString(*this, "CHttpException");
}


CAuthenticationException::CAuthenticationException(const std::string& message,
                                                   int status,
                                                   const std::string& reason,
                                                   const std::string& method,
                                                   const std::string& url)
    : CHttpException(message, status, reason, method, url) {
}

std::string CAuthenticationException::ToString() const {
    return CHttpExceptionToString(*this, "CAuthenticationException");
}


CFileNotFoundException::CFileNotFoundException(const std::string& message,
                                               const CPath& path)
    : CStorageException(message),
    path_(path) {
}

std::string CFileNotFoundException::ToString() const {
    std::ostringstream tmp;
    tmp << "CFileNotFoundException: " << what() << " " << path();
    return tmp.str();
}



CInvalidFileTypeException::CInvalidFileTypeException(const CPath& path,
                                                     bool blob_expected)
    : path_(path),
      blob_expected_(blob_expected) {
}

std::string CInvalidFileTypeException::ToString() const {
    std::ostringstream tmp;
    tmp << "CInvalidFileTypeException: " << what() << " " << path()
            << ": expected " << (blob_expected() ? "blob" : "folder");
    return tmp.str();
}


CRetriableException::CRetriableException(std::exception_ptr p_cause,
                                         std::chrono::milliseconds delay) :
        CStorageException("Wrapped to be retried", p_cause),
        delay_(delay) {
}

std::string CRetriableException::ToString() const {
    std::ostringstream tmp;
    tmp << "CRetriableException: " << what();
    return tmp.str();
}


// Below are some utility functions to pretty log exceptions

static void TrimTrailingNulls(std::string *p_string) {
    for (auto it = p_string->begin(); it != p_string->end(); ++it) {
        if (*it == '\0') {
            p_string->erase(it, p_string->end());
            break;
        }
    }
    boost::algorithm::trim(*p_string);
}

// How to display an http_exception from cpprest:
static std::ostream& operator<<(std::ostream& stream,  // NOLINT
                                const web::http::http_exception& e) {
    stream << "http_exception: " << e.what();
    if (e.error_code().value() != 0) {
        // may have leading \r\0\0\0\0....\0
        std::string msg = e.error_code().message();
        TrimTrailingNulls(&msg);
        stream << " (" << e.error_code() << ": " << msg << ")";
    }
    return stream;
}

static std::string ExceptionToString(const web::http::http_exception& e) {
    std::ostringstream tmp;
    tmp << e;
    return tmp.str();
}

static std::ostream& operator<<(std::ostream& stream,  // NOLINT
                                const std::system_error& e) {
    std::string msg = e.code().message();
    TrimTrailingNulls(&msg);  // in case
    stream << "system_error: " << e.what()
           << " (" << e.code() << ": " << msg << ")";
    return stream;
}

static std::string ExceptionToString(const std::system_error& e) {
    std::ostringstream tmp;
    tmp << e;
    return tmp.str();
}

static std::ostream& operator<<(std::ostream& stream,
                                const std::exception& se) {
    stream << typeid(se).name() << ": " << se.what();
    return stream;
}

static std::string ExceptionToString(const std::exception& se) {
    std::ostringstream tmp;
    tmp << se;
    return tmp.str();
}


/**
 * \brief Add boost detailed exception info, if available.
 */
static void AddBoostDetailedInfo(std::ostream *p_out) {
    boost::exception const *p_be =
                       boost::current_exception_cast<boost::exception const>();
    if (!p_be) {
        return;
    }
    char const * const * f = boost::get_error_info<boost::throw_file>(*p_be);
    int const * l = boost::get_error_info<boost::throw_line>(*p_be);
    char const * const * fn =
                           boost::get_error_info<boost::throw_function>(*p_be);
    if (f && l && fn) {
        *p_out << std::endl << "  ";
        *p_out << *f;
        *p_out << '(' << *l << "): ";
        *p_out << "Throw in function ";
        *p_out << *fn;
    }
}

std::string CurrentExceptionToString() {
    return ExceptionPtrToString(std::current_exception());
}

std::string ExceptionPtrToString(const std::exception_ptr p_exception) {
    if (!p_exception) {
        return "[exception_ptr=nullptr ?!]";
    }
    try {
        std::rethrow_exception(p_exception);
    }
    catch (web::http::http_exception& e) {
        return ExceptionToString(e);
    }
    catch (std::system_error& e) {
        return ExceptionToString(e);
    }
    catch (CStorageException& cse) {
        std::ostringstream tmp;
        tmp << cse.ToString();
        AddBoostDetailedInfo(&tmp);

        // Add cause information, if exists:
        if (cse.cause()) {
            tmp << std::endl << "Caused by: ";
            tmp << ExceptionPtrToString(cse.cause());
        }
        return tmp.str();
    }
    catch (std::exception& se) {
        std::ostringstream tmp;
        tmp << ExceptionToString(se);
        AddBoostDetailedInfo(&tmp);
        return tmp.str();
    }
    catch (...) {
        return "Unknown exception object ?!";
    }
    return "should not happen";
}

}  // namespace pcs_api
